### DISCUSSION TITLE
Is it possible to use the library with pytest.fixture and MagicMock?

### DISCUSSION THREAD

**Discussion by samuelint** [zhukovgreen/pytest-when#31] [Q&A] [answered]
Hi! I see a lot of potential in this library. 

I have a use case scenario where I need to use the `when` for the dependency of a tested class. 
The following code does not work. I have the error `TypeError: 'module' object is not callable`, which make sense if the library only support module path. 

```python
class TestMyInstance:
    @pytest.fixture
    def agent(self):
        agent = MagicMock(spec=CompiledGraph)
        when(agent, "invoke").called_with(
            {"messages": [{"role": "user", "content": "hello"}]},
        ).then_return({"messages": [AIMessage(id="a", content="Hello world!")]})

        return agent

    @pytest.fixture
    def instance(self, agent):
        return MyInstance.from_agent(
            agent=agent
        )

    def test_involke(self, instance):
        result = instance.invoke(...)

       assert result == "my expected result"

```

Is there a way to make the library working for this use case?
What would be the recommended way to use the library when mocking a class dependency?

**Comment by zhukovgreen** [comment]
Hi @samuelint , thanks for the feedback and your question. I'll take a look to it tomorrow and get back here.

**Comment by zhukovgreen** [comment, marked as answer]
Next time please care about others' time and give a reproducible code that I can run locally without any extra editions.

From what I see, there is no problem with pytest-when in this case and it can support your use case without any limitations. 

The core problem in your code was wrong call to when fixture and fixtures compositions. when fixture, identically to unittest.mock.patch.object, takes Klass as target 
(type hints are (cls: _TargetClsType, method: str)). So instead of when(agent, ...) you need to have when(CompiledGraph, ...). 

Then you had issue with fixtures composition in your fixtures. 

Working solution
```python
from typing import NamedTuple

import pytest

from pytest_mock import MockerFixture

from pytest_when.when import When


class AIMessage(NamedTuple):
    id: str
    content: str


class CompiledGraph:
    @classmethod
    def from_agent(
        cls,
        agent: "CompiledGraph",
    ) -> "MyInstance":
        return cls()

    def invoke(self, messages: list[AIMessage]): ...


class TestMyInstance:
    @pytest.fixture()
    def agent(self, when: When, mocker: MockerFixture):
        agent = mocker.MagicMock(spec=CompiledGraph)
        when(CompiledGraph, "invoke").called_with(
            messages=[{"role": "user", "content": "hello"}]
        ).then_return(
            {"messages": [AIMessage(id="a", content="Hello world!")]}
        )

        return agent

    @pytest.fixture()
    def instance(self, agent):
        return CompiledGraph.from_agent(agent=agent)

    def test_involke(self, instance):
        result = instance.invoke(
            messages=[{"role": "user", "content": "hello"}]
        )

        assert result == {
            "messages": [AIMessage(id="a", content="Hello world!")]
        }
```

**Comment by samuelint** [comment]
Thanks for taking time to find a solution. However, I was wondering if it was possible to pass the magic mock instance to the when function. Instead of declaring the class type (CompiledGraph in this case).

```python
agent = MagicMock(spec=CompiledGraph)
when(agent, "invoke")
```

From what I understand, it's not possible :)

---
This library resembles a lot like https://www.npmjs.com/package/jest-when (typescript/javascript), where it's possible to pass instance of the mocked object. I think being able to do the same would be a great add-on :) That's something we often see in other languages testing framework, but does not seem to be the case in python.

**Reply by zhukovgreen** [reply]
ok, I understand your request better now. The point is that pytest-when uses autospecing [1] under the hood when patching the object. Autospecing does not allow to have a target that is already mocked. This is basically the issue. 


[1] https://docs.python.org/3/library/unittest.mock.html#auto-speccing
